x/ccv/consumer/keeper/validators.go

```diff
@@ -100,13 +102,74 @@ func (k Keeper) ValidatorByConsAddr(sdk.Context, sdk.ConsAddress) stakingtypes.V
 	return stakingtypes.Validator{}
 }
 
+func (k Keeper) SetLargestSoftOptOutValidatorPower(ctx sdk.Context, power uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.SoftOptOutThresholdPowerKey(), sdk.Uint64ToBigEndian(power))
+}
+
+// UpdateLargestSoftOptOutValidatorPower sets the largest validator power that is allowed to soft opt out
+// This is the largest validator power such that the sum of the power of all validators with a lower or equal power
+// is less than 5% of the total power of all validators
+func (k Keeper) UpdateLargestSoftOptOutValidatorPower(ctx sdk.Context) {
+	// get all validators
+	valset := k.GetAllCCValidator(ctx)
+
+	// sort validators by power ascending
+	sort.Slice(valset, func(i, j int) bool {
+		return valset[i].Power < valset[j].Power
+	})
+
+	// get total power in set
+	totalPower := sdk.ZeroDec()
+	for _, val := range valset {
+		totalPower = totalPower.Add(sdk.NewDecFromInt(sdk.NewInt(val.Power)))
+	}
+
+	// get power of the biggest validator who is allowed to soft opt out
+	powerSum := sdk.ZeroDec()
+	for _, val := range valset {
+		powerSum = powerSum.Add(sdk.NewDecFromInt(sdk.NewInt(val.Power)))
+		// if powerSum / totalPower > SoftOptOutThreshold
+		if powerSum.Quo(totalPower).GT(sdk.MustNewDecFromStr(k.GetSoftOptOutThreshold(ctx))) {
+			// set largestSoftOptOutValidatorPower to the power of this validator
+			k.SetLargestSoftOptOutValidatorPower(ctx, uint64(val.Power))
+			k.Logger(ctx).Info("largest soft opt out validator power updated", "power", val.Power)
+			return
+		}
+	}
+	// This will be hit if the SoftOptOutThreshold param is equal to 1
+	panic("unreachable")
+}
+
+// GetSoftOptOutThresholdPower returns the largest validator power that is allowed to soft opt out
+func (k Keeper) GetSoftOptOutThresholdPower(ctx sdk.Context) int64 {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.SoftOptOutThresholdPowerKey())
+	if bz == nil {
+		return 0
+	}
+	return int64(binary.BigEndian.Uint64(bz))
+}
+
 // Slash queues a slashing request for the the provider chain
 // All queued slashing requests will be cleared in EndBlock
 func (k Keeper) Slash(ctx sdk.Context, addr sdk.ConsAddress, infractionHeight, power int64, _ sdk.Dec, infraction stakingtypes.InfractionType) {
 	if infraction == stakingtypes.InfractionEmpty {
 		return
 	}
 
+	// if this is a downtime infraction and the validator is allowed to
+	// soft opt out, do not queue a slash packet
+	if infraction == stakingtypes.Downtime {
```

Comment:
The `infractionHeight` for downtime misbehavior is the current height minus two, see https://github.com/cosmos/cosmos-sdk/blob/b05b6fe651514c11af3d4160f7c75fbaad92d5db/x/slashing/keeper/infractions.go#L89. This means that `SoftOptOutThresholdPower` should contain info from two height ago.

Reply:
@mpoke even with this order of begin blockers, wouldn't the soft opt out logic still be off by 1 block? (not 2)

Reply:
I don't think so. Consider the following example.

- At height 9, in ccv.EndBlock the consumer updates the valset in the CCV module using the powers received from the provider at height 9, see https://github.com/cosmos/interchain-security/blob/afcc4f558444e377d729a46cd4eaf2c278da5573/x/ccv/consumer/module.go#L202
- At height 10, in ccv.BeginBlock the consumer updates smallest validator power that cannot opt out using the valset in the CCV module that was updated at height 9, see https://github.com/cosmos/interchain-security/blob/afcc4f558444e377d729a46cd4eaf2c278da5573/x/ccv/consumer/module.go#L153
- At height 11, in slashing.BeginBlock the consumer calls ccv.Slash for a downtime infraction, which means that the infraction height is 9 (see https://github.com/cosmos/cosmos-sdk/blob/c0fe4f7da17b7ec17d9bea6fcb57b4644f044b7a/x/slashing/keeper/infractions.go#L85). Then, `ccv.Slash` compares the validator's power at height 9 with the smallest power that was updated at height 10 using powers from height 9.

Reply:
Ah makes sense, neat how that works out 👍